lib/archethic/mining/pending_transaction_validation.ex

```diff
@@ -825,49 +832,129 @@ defmodule Archethic.Mining.PendingTransactionValidation do
   end
 
   defp verify_token_resupply(tx, %{"token_reference" => token_ref}) do
-    with {:ok, token_address} <- Base.decode16(token_ref, case: :mixed),
-         # verify same chain
-         {:ok, genesis_address} <- fetch_previous_tx_genesis_address(tx),
-         storage_nodes <-
-           Election.chain_storage_nodes(token_address, P2P.authorized_and_available_nodes()),
-         {:ok, ^genesis_address} <-
-           TransactionChain.fetch_genesis_address(token_address, storage_nodes),
-         # verify token_reference
-         {:ok,
-          %Transaction{
-            data: %TransactionData{
-              content: content
-            }
-          }} <- TransactionChain.fetch_transaction(token_address, storage_nodes),
+    # strict because there was a json schema validation before
+    token_address = Base.decode16!(token_ref, case: :mixed)
+
+    storage_nodes =
+      Election.chain_storage_nodes(token_address, P2P.authorized_and_available_nodes())
+
+    # fetch in parallel the data we need
+    tasks = [
+      Task.Supervisor.async_nolink(TaskSupervisor, fn ->
+        fetch_previous_tx_genesis_address(tx)
+      end),
+      Task.Supervisor.async_nolink(TaskSupervisor, fn ->
+        TransactionChain.fetch_genesis_address(token_address, storage_nodes)
+      end),
+      Task.Supervisor.async_nolink(TaskSupervisor, fn ->
+        TransactionChain.fetch_transaction(token_address, storage_nodes)
+      end)
+    ]
+
+    # Shut down the tasks that did not reply nor exit
+    results =
+      [tx_genesis_result, ref_genesis_result, ref_tx_result] =
+      Task.yield_many(tasks)
+      |> Enum.map(fn {task, res} ->
+        res || Task.shutdown(task, :brutal_kill)
+      end)
+
+    with 0 <- Enum.count(results, &(&1 == nil)),
```

Comment:
I think you don't need this since any failed task will return nil, so the next intruction in the `with` will fail and you'll get a `nil` in the else. So you can just pattern match in the else on `nil`